### DISCUSSION TITLE
Find ip address

### DISCUSSION THREAD

**Discussion by CR-NitzanMilchin** [utmapp/UTM#5766] [macOS]
is there anyway to find the ip address of a macos vm?
using ip-address from utmctl dosn't work. it says "Operation not supported by the backend"

thank you.

**Comment by delan** [comment]
i’ve found a workaround, but it’s annoying. we can query ip addresses over serial as follows:

- add a serial port in the virtual machine config
- in the guest, append two lines to /etc/gettytab (based on [this](https://superuser.com/a/1062709))
  - `std.virtio:\`
  - `	:np:im=\r\n:sp#115200:tt=vt220:`
- in the guest, run `while :; do sudo /usr/libexec/getty std.virtio cu.virtio; done` (based on [this](https://superuser.com/a/1062709))
  - better to set up a LaunchDaemon in the long term, but this was good enough for quick testing
- log in over serial, then run `ifconfig` to get all of the ip addresses

how to log in over serial depends on the **Mode** of the serial port in the virtual machine config:

- **Built-in Terminal** → use the terminal window that opens on the host (`utmctl attach` is not implemented yet)
- [**Pseudo-TTY Device**](https://docs.getutm.app/advanced/serial/) → can be automated, see below

you can log in manually with `picocom` (`brew install picocom`): [^1]

[on the host] `picocom -b115200 --omap lfcrlf ...`

i have no idea how to automate querying ip addresses with `picocom`, but you could try something like this:

[on the host] `dev=...; echo '[username]' > $dev; sleep 0.5; echo '[password]' > $dev; echo 'echo @@@START@@@;ifconfig;echo @@@EOF@@@;exit' > $dev; cat $dev | sed '1,/@@@START@@@/d;/@@@EOF@@@/q' | egrep '^ *inet6? '`

<img width="804" height="630" alt="image" src="https://github.com/user-attachments/assets/fcb607de-5bb2-4a66-89ac-c1ef2f875ffe" />

---

i also tried unsuccessfully to log in with `cu` (based on [this](https://jasonmurray.org/posts/2020/serialconsolemacos/)).

for some reason, `cu` refuses to even connect until you `sudo chown $(whoami) /var/spool/uucp`:

```
$ cu -d -115200 -l ...
cu: creat during lock (/var/spool/uucp/TMP000000b2fd in /Users/shuppy as uid 503): Permission denied
cu: /dev/ttys001: Line in use

$ sudo cu -d -115200 -l ...
cu: open (/dev/ttys001): Permission denied
cu: /dev/ttys001: Line in use

$ ls -ld /var/spool/uucp ...
crw--w----  1 shuppy  tty    0x10000001 Nov 26 22:32 /dev/ttys001
drwxr-xr-x  2 _uucp   wheel          64 Nov 26 22:44 /var/spool/uucp
```

but even then, nothing gets through if i connect to getty with `cu -d -115200 -l ...`.

probably i’m using `cu` wrong.

[^1]: from now on, every time you see `...`, type `$(osascript -e 'tell application "UTM"' -e 'set vm to virtual machine named "my virtual machine"' -e 'address of serial port of vm' -e 'end tell')`
